### ISSUE TITLE
Mesh rendering results in a white image.

### ISSUE THREAD

**Issue by akshaygadipatil** [facebookresearch/pytorch3d#1046]
If you do not know the root cause of the problem / bug, and wish someone to help you, please
post according to this template:

## 🐛 Bugs / Unexpected behaviors

When trying to render a triangle mesh using PerspectiveCameras, the resulting image is white. Can't figure why this is the case.

NOTE: Please look at the existing list of Issues tagged with the label ['bug`](https://github.com/facebookresearch/pytorch3d/issues?q=label%3Abug). **Only open a new issue if this bug has not already been reported. If an issue already exists, please comment there instead.**.

## Instructions To Reproduce the Issue:

Please run the below code with the the [linked](https://drive.google.com/file/d/1CzhTLwnymmJkWlc-qzmqZm4JpOOgah4n/view?usp=sharing) '.obj' mesh file to reproduce the issue.

Please include the following (depending on what the issue is):

1. Any changes you made (`git diff`) or code you wrote
```

`import os
import matplotlib.pyplot as plt
import numpy as np
import torch.cuda
from PIL import Image
import cv2
import open3d as o3d
import math
from pytorch3d.io import load_objs_as_meshes, load_obj

from pytorch3d.vis.plotly_vis import plot_batch_individually
from pytorch3d.structures import Pointclouds
from pytorch3d.renderer.cameras import PerspectiveCameras
from pytorch3d.renderer import (
    TexturesVertex, BlendParams,
    FoVOrthographicCameras, look_at_view_transform,
    RasterizationSettings, PointsRasterizationSettings,
    PointsRasterizer, PointsRenderer, AlphaCompositor,
    Materials, PointLights,
    RasterizationSettings, MeshRenderer, MeshRasterizer,
    HardFlatShader, SoftPhongShader
)
from pytorch3d.structures import Meshes


def render_mesh(img_path, mesh_object, verts, device):

    img_height = 479
    img_width = 640

    ###################### Using PyTorch3D ################################
    R = torch.from_numpy(np.expand_dims(np.eye(3), axis=0)).float()#.to(device)
    T = torch.from_numpy(np.expand_dims(np.zeros(3), axis=0)).float()#.to(device)

    M = torch.eye(3).to(R.device)
    M[0, 0] = -1.0
    M[1, 1] = -1.0
    # convert R, T from your to Py3D coordinate system
    R = M @ R @ M.inverse()
    # T = (M @ torch.transpose(T, 0, 1))
    cameras = PerspectiveCameras(
        focal_length=((699, 699),),
        principal_point=((img_width/2, img_height/2),),
        R=R,
        T=T,
        device=device,
        in_ndc=False,
        image_size=((img_height, img_width),),
    )
    verts = verts.unsqueeze(0).float().to(device)
    # xy = cameras.transform_points(verts)[:, :, :2]

    raster_settings = RasterizationSettings(
        image_size=(img_height, img_width),
        blur_radius= 0, #np.log(1. / 1e-4 - 1.) * blend_params.sigma, #0.0, # make it zero for getting good flat shaded renderings
        faces_per_pixel=1,
        cull_backfaces=True
    )
    # Initialize rasterizer by using a MeshRasterizer class
    rasterizer = MeshRasterizer(
        cameras=cameras,
        raster_settings=raster_settings
    )
    lights = PointLights(location=torch.tensor([-1.0, 1.0, -2.0], device=device)[None], device=device)

    flat_shader = HardFlatShader(
        device=device, 
        cameras=cameras, 
        lights=lights, 
        blend_params=BlendParams(background_color=(0.0, 0.0, 1.0)))

    phong_shader = SoftPhongShader(
        device=device,
        cameras=cameras,
        lights=lights,
        blend_params=BlendParams(background_color=(0.0, 0.0, 1.0)),
    )

    flat_renderer = MeshRenderer(rasterizer, flat_shader)
    phong_renderer = MeshRenderer(rasterizer, phong_shader)

    images = flat_renderer(mesh_object)
    np_image = images.squeeze(0).cpu().detach().numpy()
    rgb_image = np_image[:,:,:3]
    alpha_channel = np_image[:,:,3]

    plt.imshow(alpha_channel)
    plt.show()
    exit()
    print(images)
    ######################### Using PyTorch3D ###################################


if __name__ == '__main__':

    mesh_obj_path = '../bcf1597b63e7712da04260cfca128d26/initial_cad_models/tv.obj'
    if torch.cuda.is_available():
        device = torch.device("cuda:0")
        torch.cuda.set_device(device)
    else:
        device = torch.device("cpu")

    verts, faces, aux = load_obj(mesh_obj_path, load_textures=True,
                                 create_texture_atlas=False, texture_atlas_size=4, texture_wrap="repeat")

    # verts_rgb = torch.ones_like(verts)[None]  # (1, V, 3)
    verts_rgb = torch.rand_like(verts)[None]
    textures = TexturesVertex(verts_features=verts_rgb.to(device))
    mesh_object = Meshes(
        verts=[verts.to(device)],
        faces=[faces.verts_idx.to(device)],
        textures=textures)
    render_mesh(mesh_object, verts, device)
`
```
2. The exact command(s) you ran: python script.py (where 'script.py' is the name of the script containing the above code)
3. What you observed (including the full logs):
```
Just an empty image of the specified size (WxH = 640x479). No logs to show or print.
```

Please also simplify the steps as much as possible so they do not require additional resources to
   run, such as a private dataset.

**Comment by bottler** [comment]
I extracted the parts of your code that load the mesh and setup the cameras. Then a call to `plot_batch_individually([mesh_object, cameras])` shows that the camera is pointing away from the object. That's why you can't see anything.

**Comment by akshaygadipatil** [comment]
Thank you, @bottler. I tried to visualize the camera+mesh setup using `plot_batch_individually([mesh_object, cameras])`, but I don't see any plot. The above line is executed though. Am I missing some command to display the plot (like `plt.show(),` that doesn't work either btw)? Appreciate your prompt response.

**Comment by bottler** [comment]
I'm not sure. You can't run it inside a function (e.g. inside render_mesh). I had to rearrange the code so that mesh_object and cameras were global in the notebook. Then I ran the plot_batch_individually on its own in a single cell.

**Comment by akshaygadipatil** [comment]
Can you drop your code here? I moved `mesh_object` and `cameras` outside of my defined function (i.e., render_mesh), so that they are now global variables. I tried to run `plot_batch_individually([mesh_object, cameras])`, but I don't see any plot. It is successfully executed, but no plot to see. It's unlikely, but do you think it could be a potential visualization bug with pytorch3d.vis.plotly_vis?

**Comment by bottler** [comment]
Are you in a notebook in a browser environment? What is your setup?

**Comment by akshaygadipatil** [comment]
I am running locally on a GPU, with PyCharm IDE.

**Comment by akshaygadipatil** [comment]
I managed to visualize the mesh+camera setting in Jupyter notebook
![Screenshot from 2022-01-18 10-59-14](https://user-images.githubusercontent.com/22264995/150001499-1ce9d8dc-7241-406f-9786-ae0e50d7adef.png).

But that's an inconvenience. Can we not visualize it in PyCharm IDE? Do I have to open a Jupyter notebook everytime to visualize such setups?

**Comment by bottler** [comment]
I have no idea about PyCharm / standalone visualization. But you should be able to fix your original problem now by redirecting the camera - e.g. by changing R[2,2] from 1 to -1.

**Comment by akshaygadipatil** [comment]
Yes, redirecting the camera fixes the problem. I do have something else (related to this open issue) to ask, but I am getting visualizations first to describe it.  I will get back to you with my question soon. So, for now, I am keeping the issue open. Thanks.

**Comment by akshaygadipatil** [comment]
My question is regarding Camera Viewpoint for rendering. So, this may not be the best question under the title of this issue. Let me know if you want me to open a new issue for this. 

Here is the situation:
- I have an RGB image, whose camera pose information is unknown.
- I have a pair of CAD models that try to represent the image in 3D. Ignore whether or not the CAD models are similarly oriented as the objects in the image. I want to render the pair of CAD models from the same viewpoint as the RGB image.
- So far, I have been able to render the pair by using the following camera parameters
    ```
    R = torch.from_numpy(np.expand_dims(np.eye(3), axis=0)).float()#.to(device)
    T = torch.from_numpy(np.expand_dims(np.zeros(3), axis=0)).float()#.to(device)    
    R[0,0,0] = -1
    R[0,2,2] = -1

    cameras = PerspectiveCameras(
        focal_length=((500, 500),),
        principal_point=((img_width/2, img_height/2),),
        R=R,
        T=T,
        device=device,
        in_ndc=False,
        image_size=((img_height, img_width),),
    )
```

- And the rendered image, along with the RGB image look like these

![bcf1597b63e7712da04260cfca128d26](https://user-images.githubusercontent.com/22264995/150066202-456808d7-91c2-4fe4-82cc-f43e8d122d48.jpg)

![rgb_channels](https://user-images.githubusercontent.com/22264995/150066249-33a3eadc-0f7a-4701-8eb8-f86b82696db5.png)

Is there a way to get the exact view-rendering as the input RGB image?

**Comment by bottler** [comment]
That's a research question rather than a question about PyTorch3D, and we can't help here. It's a bit like what happens in the camera position optimization tutorial https://pytorch3d.org/tutorials/camera_position_optimization_with_differentiable_rendering .

**Comment by akshaygadipatil** [comment]
I understand. But can you tell me if there is a way I can adjust the camera position explicitly using `PerspectiveCameras` (similar to the `eye` argument in `look_at_transform`)?

**Comment by bottler** [comment]
I don't totally understand. The camera position is set by the R and T inputs. You can change them. There are also ways to determine them (e.g. in PyTorch3D: [efficient_pnp](https://github.com/facebookresearch/pytorch3d/blob/v0.6.1/pytorch3d/ops/perspective_n_points.py#L303) to try to match points in the image to known points in the mesh, [rotate_on_spot](https://github.com/facebookresearch/pytorch3d/blob/v0.6.1/pytorch3d/renderer/camera_utils.py#L71) to preserve a location but change direction). If by `eye` you mean to change the camera location without changing the direction, that's just changing T without changing R.
